File: backend/controllers/teacher.controller.js
const db = require('../config/database');
const { sendSuccess } = require('../utils/response');

const ollamaService = require('../services/ollamaService');

const pdfParse = require('pdf-parse');
const mammoth = require('mammoth');

const getSubjects = async (req, res) => sendSuccess(res, 200, 'Ok', ['Java Programming', 'Database Systems', 'Software Engineering']);

const generateQuestionsFromPrompt = async (req, res, next) => {
  try {
    const { prompt, subject, questionType, difficulty, numberOfQuestions } = req.body;
    const count = numberOfQuestions || 5;
    
    // Treat the prompt as the context/syllabus for Ollama
    const questions = await ollamaService.generateQuestions(
      prompt || 'Generate general questions for this subject',
      count,
      subject || 'General Knowledge',
      difficulty || 'Medium',
      questionType || 'MCQ'
    );
    
    sendSuccess(res, 200, 'Questions generated successfully', { questions });
  } catch (error) {
    next(error);
  }
};

const generateQuestionsFromFile = async (req, res, next) => {
  try {
    if (!req.file) {
      return res.status(400).json({ success: false, message: 'No file uploaded' });
    }

    const { subject, questionType, difficulty, numberOfQuestions } = req.body;
    const count = numberOfQuestions || 5;
    let extractedText = '';

    // Parse file based on mimetype
    if (req.file.mimetype === 'application/pdf') {
      const data = await pdfParse(req.file.buffer);
      extractedText = data.text;
    } else if (req.file.mimetype === 'application/vnd.openxmlformats-officedocument.wordprocessingml.document') {
      const result = await mammoth.extractRawText({ buffer: req.file.buffer });
      extractedText = result.value;
    } else {
      return res.status(400).json({ success: false, message: 'Unsupported file type. Please upload a PDF or DOCX file.' });
    }

    if (!extractedText.trim()) {
      return res.status(400).json({ success: false, message: 'Could not extract text from the file. Please ensure it contains readable text.' });
    }

    // Limit text to roughly 2000 words (~12,000 chars) to avoid overloading Ollama
    const maxChars = 12000;
    if (extractedText.length > maxChars) {
      extractedText = extractedText.substring(0, maxChars) + '...';
    }

    const questions = await ollamaService.generateQuestions(
      extractedText,
      count,
      subject || 'General Knowledge',
      difficulty || 'Medium',
      questionType || 'MCQ'
    );

    if (!questions || questions.length === 0) {
      return res.status(500).json({ success: false, message: 'AI failed to generate questions in the correct format. Please try again.' });
    }

    sendSuccess(res, 200, 'Questions generated successfully from document', { questions });
  } catch (error) {
    next(error);
  }
};

module.exports = {
  getSubjects,
  generateQuestionsFromPrompt,
  generateQuestionsFromFile
};

File: backend-spring/src/main/java/com/examify/backend/controller/TeacherController.java
package com.examify.backend.controller;

import com.examify.backend.dto.ApiResponse;
import com.examify.backend.dto.TeacherDto;
import com.examify.backend.exception.ApiException;
import com.examify.backend.service.DocumentParsingService;
import com.examify.backend.service.OllamaService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/teacher")
@RequiredArgsConstructor
public class TeacherController {

    private final OllamaService ollamaService;
    private final DocumentParsingService documentParsingService;

    @GetMapping("/subjects")
    public ResponseEntity<ApiResponse<List<String>>> getSubjects() {
        return ResponseEntity.ok(ApiResponse.success("Ok", List.of("Java Programming", "Database Systems", "Software Engineering")));
    }

    @PostMapping("/generate-questions-from-prompt")
    public ResponseEntity<ApiResponse<Map<String, Object>>> generateQuestionsFromPrompt(@RequestBody TeacherDto.GeneratePromptRequest request) {
        int count = request.getNumberOfQuestions() != null ? request.getNumberOfQuestions() : 5;
        String prompt = request.getPrompt() != null ? request.getPrompt() : "Generate general questions for this subject";
        String subject = request.getSubject() != null ? request.getSubject() : "General Knowledge";
        String diff = request.getDifficulty() != null ? request.getDifficulty() : "Medium";
        String qType = request.getQuestionType() != null ? request.getQuestionType() : "MCQ";

        List<TeacherDto.GeneratedQuestion> questions = ollamaService.generateQuestions(prompt, count, subject, diff, qType, null);
        
        Map<String, Object> data = new HashMap<>();
        data.put("questions", questions);
        return ResponseEntity.ok(ApiResponse.success("Questions generated successfully", data));
    }

    @PostMapping(value = "/generate-questions-from-file", consumes = {"multipart/form-data"})
    public ResponseEntity<ApiResponse<Map<String, Object>>> generateQuestionsFromFile(
            @RequestParam("document") MultipartFile document,
            @RequestParam(value = "subject", defaultValue = "General Knowledge") String subject,
            @RequestParam(value = "questionType", defaultValue = "MCQ") String questionType,
            @RequestParam(value = "difficulty", defaultValue = "Medium") String difficulty,
            @RequestParam(value = "numberOfQuestions", defaultValue = "5") Integer numberOfQuestions) {
        
        if (document.isEmpty()) {
            throw new ApiException(HttpStatus.BAD_REQUEST, "No file uploaded");
        }

        try {
            String extractedText = documentParsingService.extractText(document);
            
            if (extractedText == null || extractedText.trim().isEmpty()) {
                throw new ApiException(HttpStatus.BAD_REQUEST, "Could not extract text from the file.");
            }

            int maxChars = 12000;
            if (extractedText.length() > maxChars) {
                extractedText = extractedText.substring(0, maxChars) + "...";
            }

            List<TeacherDto.GeneratedQuestion> questions = ollamaService.generateQuestions(
                    extractedText, numberOfQuestions, subject, difficulty, questionType, null);

            if (questions == null || questions.isEmpty()) {
                throw new ApiException(HttpStatus.INTERNAL_SERVER_ERROR, "AI failed to generate questions in the correct format.");
            }

            Map<String, Object> data = new HashMap<>();
            data.put("questions", questions);
            return ResponseEntity.ok(ApiResponse.success("Questions generated successfully from document", data));
            
        } catch (ApiException e) {
            throw e;
        } catch (Exception e) {
            throw new ApiException(HttpStatus.INTERNAL_SERVER_ERROR, "Error parsing file: " + e.getMessage());
        }
    }
}
